Win32.xs

```diff
@@ -1682,6 +1693,248 @@ XS(w32_IsDeveloperModeEnabled)
+
+    /* Call WinHttpGetProxyForUrl with our target URL. If auto-proxy succeeds,
+     * then set the proxy info on the request handle. If auto-proxy fails,
+     * ignore the error and attempt to send the HTTP request directly to the
+     * target server (using the default WINHTTP_ACCESS_TYPE_NO_PROXY
+     * configuration, which the request handle will inherit from the session).
+     */
+    if (hRequest) {
+        ZeroMemory(&AutoProxyOptions, sizeof(AutoProxyOptions));
+        ZeroMemory(&ProxyInfo, sizeof(ProxyInfo));
+        AutoProxyOptions.dwFlags = WINHTTP_AUTOPROXY_AUTO_DETECT;
+        AutoProxyOptions.dwAutoDetectFlags =
+                                    WINHTTP_AUTO_DETECT_TYPE_DHCP |
+                                    WINHTTP_AUTO_DETECT_TYPE_DNS_A;
+        AutoProxyOptions.fAutoLogonIfChallenged = TRUE;
+
+        if(WinHttpGetProxyForUrl(hSession,
+                                url,
+                                &AutoProxyOptions,
+                                &ProxyInfo)) {
+            if(!WinHttpSetOption(hRequest,
+                                WINHTTP_OPTION_PROXY,
+                                &ProxyInfo,
+                                cbProxyInfoSize)) {
+                bAborted = TRUE;
+                Perl_warn(aTHX_ "Win32::HttpGetFile: setting proxy options failed");
+            }
+            Safefree(ProxyInfo.lpszProxy);
+            Safefree(ProxyInfo.lpszProxyBypass);
+        }
+    }
+
+    /* Send a request. */
+    if (hRequest && !bAborted)
+        bResults = WinHttpSendRequest(hRequest,
+                                      WINHTTP_NO_ADDITIONAL_HEADERS,
+                                      0,
+                                      WINHTTP_NO_REQUEST_DATA,
+                                      0,
+                                      0,
+                                      0);
+
+    /* End the request. */
+    if (bResults)
+        bResults = WinHttpReceiveResponse(hRequest, NULL);
+
+    /* Create output file for download. */
+    if (bResults) {
+        hOut = CreateFileW(file,
+                           GENERIC_WRITE,
+                           FILE_SHARE_READ | FILE_SHARE_WRITE,
+                           NULL,
+                           CREATE_ALWAYS,
+                           FILE_ATTRIBUTE_NORMAL,
+                           NULL);
+
+        if (!hOut || hOut == INVALID_HANDLE_VALUE)
+            bFileError = TRUE;
+    }
+
+    /* Keep checking for data until there is nothing left. */
+    if (!bFileError && bResults) {
+        do {
+            /* Check for available data. */
+            dwSize = 0;
+            if (!WinHttpQueryDataAvailable(hRequest, &dwSize)) {
+                bAborted = TRUE;
+                break;
+            }
+
+            /* No more available data. */
+            if (!dwSize)
+                break;
+
+            /* Allocate space for the buffer. */
+            New(0, pszOutBuffer, dwSize + 1, char);
```

Comment:
It's probably a better idea to have a fixed size buffer and then call `WinHttpReadData` in a loop instead of allocating it for each chunk.

Reply:
> It's probably a better idea to have a fixed size buffer and then call `WinHttpReadData` in a loop instead of allocating it for each chunk.

Hmm.  We don't know how big a buffer we need until we've called WinHttpQueryDataAvailable.  Or we potentially have to do multiple reads to get all the available data, which makes the logic more complex and harder to get right.

Reply:
BTW, I noticed that some of the variables could have a narrower scope. C89 allows declaring variables at the beginning of any block, it doesn't have to be the beginning of a function.